File: benchmark-scripts/prime-factors.js
/*
    modified from https://stackoverflow.com/questions/39899072/how-can-i-find-the-prime-factors-of-an-integer-in-javascript
*/

function getPrimeFactors(integer) {
    const primeArray = [];
    let isPrime;

    // Find divisors starting with 2
    for (let i = 2; i <= integer; i++) {
        if (integer % i !== 0) continue;

        // Check if the divisor is a prime number
        for (let j = 2; j <= i / 2; j++) {
            isPrime = i % j !== 0;
        }

        if (!isPrime) continue;
        // if the divisor is prime, divide integer with the number and store it in the array
        integer /= i
        primeArray.push(i);
    }

    return primeArray;
}

let maxFactors = 0;
let maxFactorsNum = 0;

function benchit() {
    maxFactors = 0;
    maxFactorsNum = 0;
    for (let i = 1; i < 9450; i++) {
        const factors = getPrimeFactors(i);
        if (factors.length > maxFactors) {
            maxFactors = factors.length;
            maxFactorsNum = i;
        }
    }
}

File: benchmark-scripts/zig/prime-factors.zig
const std = @import("std");

// END IMPORTS

var primeArray: std.ArrayList(u32) = undefined;

fn getPrimeFactors(integer: u32) *std.ArrayList(u32) {
    var isPrime: bool = false;
    var mutableInteger: u32 = integer;

    var i: u32 = 2;
    while (i <= mutableInteger) : (i += 1) {
        if (mutableInteger % i != 0) {
            continue;
        }

        var j: u32 = 2;
        while (j <= i / 2) : (j += 1) {
            isPrime = i % j != 0;
        }

        if (!isPrime) {
            continue;
        }

        mutableInteger /= i;
        primeArray.append(i) catch unreachable; // or handle the error appropriately
    }

    return &primeArray;
}

var maxFactors: usize = 0;
var maxFactorsNum: usize = 0;

pub fn benchit() void {
    primeArray = std.ArrayList(u32).init(std.heap.page_allocator);
    defer primeArray.deinit();
    
    maxFactors = 0;
    maxFactorsNum = 0;
    for (1..9450) |i| {
        const factors = getPrimeFactors(@as(u32, @intCast(i)));
        
        if (factors.items.len > maxFactors) {
            maxFactors = factors.items.len;
            maxFactorsNum = i;
        }

        primeArray.clearRetainingCapacity();
    }

    std.mem.doNotOptimizeAway(maxFactors);
    std.mem.doNotOptimizeAway(maxFactorsNum);
}

